src/code_builder.rs

```diff
@@ -1395,6 +1404,79 @@ fn memory_grow<C: Config>(
     Ok(Instruction::MemoryGrow(memory_index(u, builder)?))
 }
 
+#[inline]
+fn memory_init_valid<C: Config>(
+    module: &ConfiguredModule<C>,
+    builder: &mut CodeBuilder<C>,
+) -> bool {
+    have_memory(module, builder)
+        && have_data(module, builder)
+        && module.config.bulk_memory_enabled()
+        && builder.types_on_stack(&[ValType::I32, ValType::I32, ValType::I32])
+}
+
+fn memory_init<C: Config>(
+    u: &mut Unstructured,
+    module: &ConfiguredModule<C>,
+    builder: &mut CodeBuilder<C>,
+) -> Result<Instruction> {
+    let mem = memory_index(u, builder)?;
+    let data = data_index(u, module)?;
+    builder.pop_operands(&[ValType::I32, ValType::I32, ValType::I32]);
+    Ok(Instruction::MemoryInit { mem, data })
+}
+
+#[inline]
+fn memory_fill_valid<C: Config>(
+    module: &ConfiguredModule<C>,
+    builder: &mut CodeBuilder<C>,
+) -> bool {
+    have_memory(module, builder)
+        && module.config.bulk_memory_enabled()
+        && builder.types_on_stack(&[ValType::I32, ValType::I32, ValType::I32])
+}
+
+fn memory_fill<C: Config>(
+    u: &mut Unstructured,
+    _module: &ConfiguredModule<C>,
+    builder: &mut CodeBuilder<C>,
+) -> Result<Instruction> {
+    let mem = memory_index(u, builder)?;
+    builder.pop_operands(&[ValType::I32, ValType::I32, ValType::I32]);
+    Ok(Instruction::MemoryFill(mem))
+}
+
+#[inline]
+fn memory_copy_valid<C: Config>(
+    module: &ConfiguredModule<C>,
+    builder: &mut CodeBuilder<C>,
+) -> bool {
+    memory_fill_valid(module, builder)
+}
+
+fn memory_copy<C: Config>(
+    u: &mut Unstructured,
+    _module: &ConfiguredModule<C>,
+    builder: &mut CodeBuilder<C>,
+) -> Result<Instruction> {
+    let src = memory_index(u, builder)?;
+    let dst = memory_index(u, builder)?;
+    builder.pop_operands(&[ValType::I32, ValType::I32, ValType::I32]);
+    Ok(Instruction::MemoryCopy { dst, src })
+}
+
+#[inline]
+fn data_drop_valid<C: Config>(module: &ConfiguredModule<C>, builder: &mut CodeBuilder<C>) -> bool {
+    have_data(module, builder) && module.config.bulk_memory_enabled()
+}
+fn data_drop<C: Config>(
```

Comment:
Nitpick: missing a newline here.

```suggestion

fn data_drop<C: Config>(
```